go/vt/vttablet/grpctmclient/cached_client.go

```diff
@@ -0,0 +1,340 @@
+
+	index int
+	key   string
+}
+
+// cachedConns provides a priority queue implementation for O(log n) connection
+// eviction management. It is a nearly verbatim copy of the priority queue
+// sample at https://golang.org/pkg/container/heap/#pkg-overview, with the
+// Less function changed such that connections with refs==0 get pushed to the
+// front of the queue, because those are the connections we are going to want
+// to evict.
+type cachedConns []*cachedConn
+
+var _ heap.Interface = (*cachedConns)(nil)
+
+// Len is part of the sort.Interface interface and is used by container/heap
+// functions.
+func (queue cachedConns) Len() int { return len(queue) }
+
+// Less is part of the sort.Interface interface and is used by container/heap
+// functions.
+func (queue cachedConns) Less(i, j int) bool {
+	left, right := queue[i], queue[j]
+	if left.refs == right.refs {
+		// break ties by access time.
+		// more stale connections have higher priority for removal
+		// this condition is equvalent to:
+		//		left.lastAccessTime <= right.lastAccessTime
+		return !left.lastAccessTime.After(right.lastAccessTime)
+	}
+
+	// connections with fewer refs have higher priority for removal
+	return left.refs < right.refs
+}
+
+// Swap is part of the sort.Interface interface and is used by container/heap
+// functions.
+func (queue cachedConns) Swap(i, j int) {
+	queue[i], queue[j] = queue[j], queue[i]
+	queue[i].index = i
+	queue[j].index = j
+}
+
+// Push is part of the container/heap.Interface interface.
+func (queue *cachedConns) Push(x interface{}) {
+	n := len(*queue)
+	conn := x.(*cachedConn)
+	conn.index = n
+	*queue = append(*queue, conn)
+}
+
+// Pop is part of the container/heap.Interface interface.
+func (queue *cachedConns) Pop() interface{} {
+	old := *queue
+	n := len(old)
+	conn := old[n-1]
+	old[n-1] = nil  // avoid memory leak
+	conn.index = -1 // for safety
+	*queue = old[0 : n-1]
+
+	return conn
+}
+
+type cachedConnDialer struct {
+	m            sync.RWMutex
+	conns        map[string]*cachedConn
+	qMu          sync.Mutex
+	queue        cachedConns
+	connWaitSema *sync2.Semaphore
+}
+
+var dialerStats = struct {
+	ConnReuse            *stats.Gauge
+	ConnNew              *stats.Gauge
+	DialTimeouts         *stats.Gauge
+	DialTimings          *stats.Timings
+	EvictionQueueTimings *stats.Timings
+}{
+	ConnReuse:    stats.NewGauge("tabletmanagerclient_cachedconn_reuse", "number of times a call to dial() was able to reuse an existing connection"),
+	ConnNew:      stats.NewGauge("tabletmanagerclient_cachedconn_new", "number of times a call to dial() resulted in a dialing a new grpc clientconn"),
+	DialTimeouts: stats.NewGauge("tabletmanagerclient_cachedconn_dial_timeouts", "number of context timeouts during dial()"),
+	DialTimings:  stats.NewTimings("tabletmanagerclient_cachedconn_dialtimings", "timings for various dial paths", "path", "rlock_fast", "sema_fast", "sema_poll"),
+	// TODO: add timings for heap operations (push, pop, fix)
```

Comment:
I think you want to initialize evectionqueuetimings here (or remove it from the struct)